### PR TITLE
fix(router): validate nested routes

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)

Router doesn't validate nested routes (children) which cause issues https://github.com/angular/angular/issues/12827

**What is the new behavior?**

Validate children as well.

**Does this PR introduce a breaking change?** (check one with "x")
```
[x] Yes
[ ] No
```

### CODE REVIEW

File: modules/@angular/router/src/config.ts [hunk 1]
```diff
@@ -410,4 +410,7 @@ function validateNode(route: Route): void {
     throw new Error(
         `Invalid configuration of route '${route.path}': pathMatch can only be set to 'prefix' or 'full'`);
   }
+  if (route.children) {
+    validateConfig(route.children);
```
Comment (vicb) [inline modules/@angular/router/src/config.ts]: What do you think about adding the path to the config ? (instead of printing the current segment only ?)

**Comment by vicb** [comment]
Thanks, LGTM with a few comments

**Comment by vicb** [comment]
Oh and could you add `fixes #...` in the commit message?

File: modules/@angular/router/test/config.spec.ts [hunk 2]
```diff
@@ -46,6 +46,12 @@ describe('config', () => {
               `Invalid configuration of route 'a': redirectTo and children cannot be used together`);
     });
 
+    it('should validate children', () => {
+      expect(() => validateConfig([{path: 'a', children: [{path: 'b'}]}]))
+          .toThrowError(
+              `Invalid configuration of route 'b': one of the following must be provided (component or redirectTo or children or loadChildren)`);
```
Comment (DzmitryShylovich) [inline modules/@angular/router/test/config.spec.ts]: sure

**Comment by vicb** [comment]
Great, thanks !

**Comment by vicb** [comment]
Could you please make it a `fix(...)` commit ? (we should avoid "feat" with semver rls)

**Comment by jsgoupil** [comment]
This is quite a major breaking change. It think it should have been reported on the Changelog...

This uses to work, now it needs to be changed quite a bit.

```
path: ':id',
children: [
    {
        path: 'view',
        component: CustomerViewComponent,
        resolve: {
            viewData: CustomerResolve,
        },
        canDeactivate: [CanDeactivateGuard],
        children: [
            {
                path: 'edit',
            },
            {
                path: ''
            }
        ]
    },
    {
        path: '',
        redirectTo: 'view',
        pathMatch: 'full'
    }
]
```

**Comment by DzmitryShylovich** [comment]
> This uses to work

it shouldn't work. 
```
children: [
            {
                path: 'edit',
            },
            {
                path: ''
            }
        ]
```
is invalid config. can u create a small repro?

**Comment by jsgoupil** [comment]
This was working fine with router 2.3.0.

You moved forward, I don't need this to be supported. It should just have been written in the change log... no time for a plunkr for this.

thanks.

**Comment by jsgoupil** [comment]
It even say up there

"Does this PR introduce a breaking change? (check one with "x")"

It's surprising it was not noted in the changelog. That's why I'm surprised.

**Comment by DzmitryShylovich** [comment]
Validation is not a breaking change

**Comment by jsgoupil** [comment]
You must be right then. I imagined it...

**Comment by DzmitryShylovich** [comment]
> This uses to work

this is the actual bug

**Comment by dmarginian** [comment]
This is a breaking change.  Regardless of whether or not { path: 'edit' } was valid or invalid my application was relying on it, and now it is busted.

**Comment by dmarginian** [comment]
@jsgoupil how did you work-around this?  In my case I have a intermediate router-outlet that I use for a slide-out panel.  I don't want to specify a component.

**Comment by jsgoupil** [comment]
@dmarginian I rewrote my code. I didn't ship yet so my stuff is quite variable. I didn't want to have an "edit" in the URL after all especially if the user press the Back.

This breaking change came in a patch. How lovely.

**Comment by dmarginian** [comment]
Can someone from the Angular team address this?  I am sure there is an explanation but I don't see how this should happen if semver is being followed.  Especially since the PR is marked as introducing a breaking change (not backwards compatible).

**Comment by vicb** [comment]
@dmarginian if you think you have a valid issue, please create a new GH issue and add sample code there. thanks.

**Comment by dmarginian** [comment]
Thanks @vicb, and I will.  However, I am mostly interested in the status of semver.  Everything I have read said you guys are following semver since 2.0 final.  Is that not the case for the router?

**Pull request merged**